sm_ipt/sm_ipt.c

```diff
@@ -0,0 +1,450 @@
+			sm_ipt_os_atomic_or(&ctx->free_mask[0], (uint32_t)mask);
+			return false;
+		}
+	}
+
+	SM_IPT_OS_MEMORY_BARRIER();
+
+	return true;
+}
+
+static mask_t free_mask_get(struct sm_ipt_ctx *ctx)
+{
+	mask_t mask;
+
+	mask = sm_ipt_os_atomic_get(&ctx->free_mask[0]);
+
+	if (NUM_BLOCKS > 32) {
+		mask = (uint64_t)sm_ipt_os_atomic_get(&ctx->free_mask[1]) << 32;
+	}
+
+	return mask;
+}
+
+static void free_mask_init(struct sm_ipt_ctx *ctx)
+{
+	free_mask_set(ctx, ~(mask_t)0);
+}
+
+static mask_t calc_mask(size_t blocks, size_t index)
+{
+	mask_t mask;
+
+	mask = (mask_t)1 << (NUM_BLOCKS - 1);        /* 100000000... */
+	mask = (smask_t)mask >> (blocks - 1);        /* 111000000... (e.g. blocks = 3) */
+	mask = mask >> index;                        /* 000011100... (e.g. index = 4) */
+
+	return mask;
+}
+
+void sm_ipt_alloc_tx_buf(struct sm_ipt_ctx *ctx, uint8_t **buf, size_t len)
+{
+	/* Actual allocated memory: | 32-bit size | data | padding | */
+	size_t blocks = (len + (WORD_SIZE + ctx->out.block_size - 1)) / ctx->out.block_size;
+	bool sem_taken = false;
+	mask_t cur_mask;
+	mask_t sh_mask;
+	bool unset_success;
+	mask_t mask;
+	size_t free_index;
+
+	SM_IPT_DBG("Trying to allocate %d bytes", len);
+
+	if ((blocks > NUM_BLOCKS) || (blocks == 0)) {
+		SM_IPT_ERR("Requested %d bytes, maximum is %d", len,
+				ctx->out.allocable_size - sizeof(uint32_t));
+		*buf = NULL;
+		return;
+	}
+
+	do {
+		do {
+			/* create shifted mask with bits set where `blocks` can be allocated */
+			cur_mask = free_mask_get(ctx);
+			sh_mask = cur_mask;
+			for (size_t i = 1; i < blocks; i++) {
+				sh_mask &= (sh_mask << 1);
+			}
+
+			 /* if no memory */
+			if (sh_mask == 0) {
+				/* wait for any block to be empty */
+				sm_ipt_os_take(&ctx->out_sem);
+				sem_taken = true;
+			}
+
+		} while (sh_mask == 0);
+
+		/* get first available blocks */
+		free_index = mask_clz(sh_mask);
+		/* create bit mask with blocks that will be used */
+		mask = calc_mask(blocks, free_index);
+		/* update masks */
+		unset_success = free_mask_unset(ctx, mask);
+
+		/* there is a small probability that unset will be unsuccessful */
+		if (!unset_success) {
+			/* give semaphore, because free_mask_unset may cause other thread
+			 * waiting before it reverted the changes
+			 */
+			SM_IPT_DBG("Not enough free shared memory. Waiting.");
+			sm_ipt_os_give(&ctx->out_sem);
+			sem_taken = false;
+		}
+	} while (!unset_success);
+
+	/* Give semaphore back, because there may be some other thread waiting */
+	if (sem_taken && ((cur_mask & ~mask) != 0)) {
+		sm_ipt_os_give(&ctx->out_sem);
+	}
+
+	uint32_t *mem_start = (uint32_t *)&ctx->out.shm_queue->allocable[
+								ctx->out.block_size * free_index];
+
+	mem_start[0] = blocks * ctx->out.block_size;
+
+	*buf = (uint8_t *)(&mem_start[1]);
+	SM_IPT_DBG("Allocated memory at 0x%x", (uint32_t)*buf);
+}
+
+void sm_ipt_free_tx_buf(struct sm_ipt_ctx *ctx, uint8_t *buf)
+{
+	uint32_t *mem_start = (uint32_t *)buf - 1;
```

Comment:
Why we need to subtract 1 ?

Reply:
When allocating buffer the first word is buffer size and the beginning of usable data starts at second word. Here (in free functions) The function needs to get the beginning of the buffer, not the usable data, so a decrement of one (word) is required.